Provide key transformation

remear opened the pull request:
#### Purpose

Provide key case translation on a per-request basis.

```render json: posts, each_serializer: PostSerializer, key_transform: :camel_lower```

A global config option may be set to configure the key transform. ```ActiveModelSerializers.config.key_transform = :camel_lower```

#### Changes

- Adds a `key_transform` attribute to `SerializationContext`.
- Adds a `key_transform` option to `Configuration`.
- Translates keys for documents in the `JsonApi` adapter.
- Translates keys for serialized hash in the `Json` adapter.
- Provides the following casing options:
  - `camel` - `SpecialAttribute`
  - `camel_lower` - `specialAttribute`
  - `dashed` - `special-attribute` (default as per JSON API recommendation)
  - `unaltered` - the original, unaltered key
- Key translation precedence is as follows:
  - `SerializationContext` - The `key_transform` is provided as an option via render.
  - Global config - The `key_transform is set in `ActiveModelSerializers.config.key_transform`.
  - Adapter default - The default `key_transform` specified in the adapter.

remear commented:
@rails-api/ams Thoughts on this approach?

bf4 commented:
ref: JSON API

- dasherize/kebab-case, and for namespaces https://github.com/json-api/json-api/issues/850

bf4 commented on test/adapter/json_api/links_test.rb:
```diff
@@ -40,7 +40,7 @@ def test_toplevel_links
                 }
               }
-            }).serializable_hash
+            }).serializable_hash(@options)
```
:question: I don't think this should be here. Not defined anywhere and we haven't been passing anything interesting into the renderer options

bf4 replied:
there's a bug in this right now https://github.com/rails-api/active_model_serializers/pull/1572

bf4 commented:
Looks good.  you up for benchmarking it? maybe https://github.com/bf4/book_code/blob/master/adrpo/code/chp8/assert_performance.rb or something.. can go in `test/dummy` now.. which I think does need to be renamed to `test/benchmark`

remear commented:
I added a `dashed` case option and make it the default.

Also, note that rubocop wanted a lot of changes that weren't actually within the scope of this PR. I went ahead and fixed them in a separate commit.

remear commented:
The rubocop failures I see in travis right now look like the inverse of what it wanted locally, rubocop-0.38.0.

bf4 commented:
https://github.com/rails-api/active_model_serializers/pull/1577 I have a PR to fix the JRuby's.. should be green soon

bf4 commented:
```plain
Offenses:
lib/active_model/serializer/collection_serializer.rb:17:11: C: Style/GuardClause: Use a guard clause instead of wrapping the code inside a conditional expression. (https://github.com/bbatsov/ruby-style-guide#no-nested-conditionals)
          if serializer_class.nil?
          ^^
lib/active_model/serializable_resource.rb:36:5: C: Style/Alias: Use alias instead of alias_method in a class body. (https://github.com/bbatsov/ruby-style-guide#alias-method)
    alias_method :adapter_instance, :adapter
    ^^^^^^^^^^^^
lib/active_model/serializable_resource.rb:57:5: C: Style/Alias: Use alias instead of alias_method in a class body. (https://github.com/bbatsov/ruby-style-guide#alias-method)
    alias_method :serializer_class, :serializer
    ^^^^^^^^^^^^
lib/active_model_serializers/adapter.rb:4:19: C: Style/MutableConstant: Freeze mutable objects assigned to constants.
    ADAPTER_MAP = {}
                  ^^
test/support/serialization_testing.rb:24:3: C: Style/Alias: Use alias instead of alias_method in a module body. (https://github.com/bbatsov/ruby-style-guide#alias-method)
  alias_method :with_configured_adapter, :with_adapter
  ^^^^^^^^^^^^
156 files inspected, 5 offenses detected
RuboCop failed!
```

`Style/Alias` is a recent update to rubocop that was lagging the ruby-style-guide

That guard clause needs an inline `rubocop:disable Style/GuardClause` iirc

NullVoxPopuli commented on lib/active_model_serializers/adapter/base.rb:
```diff
@@ -51,6 +51,20 @@ def include_meta(json)
         json
       end
+
+      def translate_key_casing!(value, serialization_context)
```
thoughts on writing these transforms in the style of active support inflections? https://github.com/rails-api/active_model_serializers/pull/1566

at least for the sake of examples in comments, 

maybe the key transforms could be in their own module?

```ruby
module WhateverNamespace::KeyTransforms
  module_function
  # transforms keys to UpperCamelCase or PascalCase (this is what through me down this train of thought in the first place, though, this could be because I've been in non-ruby land at work)
  #
  # @example:
  #    "some_key" => "SomeKey",
  # @see: active_support/inflector#camelize
  def camel(hash)
    hash.deep_transform_keys! { |key| key.to_s.camelize.to_sym }
  end
  ...
end
```

and then this method would become:
```ruby
return value unless serialization_context
transform = serialization_context.key_case
KeyTransforms.send(transform)
```

remear replied:
I'm not necessarily opposed to it. Are there other benefits to this approach over what's in the PR?

NullVoxPopuli replied:
it would give a place for users to add their own transformations if they wish. Rather than overwriting translate_key_casing, they write their own transform method -- without losing or having to re-implement the 3 you have. I could see this being especially useful, for APIs that have clients that don't all expect the same casing.

NullVoxPopuli commented on lib/active_model_serializers/adapter/json_api.rb:
```diff
@@ -22,11 +22,12 @@ def initialize(serializer, options = {})
+                   else
+                     failure_document
+                   end
+        translate_key_casing!(document, options[:serialization_context])
```
is this method mutating? or does it return a different hash?

like,

does `document.object_id == translate_key_casing!(document, {...}).object_id`?

remear replied:
Should be mutating.

remear commented:
@NullVoxPopuli Implemented. Let me know what you think.

remear commented:
Also note that now only the `JsonApi` adapter has the `:dashed` default key case transform. The `Json` defaults to `:unaltered`.

NullVoxPopuli commented on lib/active_model_serializers/key_transform.rb:
```diff
@@ -0,0 +1,43 @@
+    module_function :camel, :camel_lower, :dashed, :unaltered
+
+    def self.method_missing(m, *args, &block)
+      warn "\"#{m}\" is not a defined key transform. Defaulting to :unaltered."
```
nice!

NullVoxPopuli replied:
can you add a test for this functionality, to make sure we don't get infinite method_missing loops?

remear replied:
Good call. I guess I forgot to add it.

NullVoxPopuli commented on test/adapter/json_api/key_case_test.rb:
```diff
@@ -0,0 +1,323 @@
+          expected_errors_object =
+            { :errors =>
+              [
+                { :source => { :pointer => '/data/attributes/published_at' }, :detail => 'must be in the future' },
```
is published_at still supposed to be underscored here?

remear replied:
This one is tricky because we're only modifying keys at the end of the document building and this is in a value. I figured we could look more into it when the error document gets links and meta support.

NullVoxPopuli replied:
right, so... maybe it would be the responsibility of the error serializer to change the keying on the attribute?

remear replied:
It's sounding ideal.

NullVoxPopuli replied:
that could be a separate PR, unless you are feeling adventurous. :-)

bf4 replied:
only question is this:  what are valid key formats for JSON API? we should limit options to those, no?

bf4 replied:
also see `config.jsonapi_resource_type = :plural`

NullVoxPopuli commented:
can you rebase, so you get the new JRuby modifications from @bf4 ?

bf4 commented on lib/active_model/serializable_resource.rb:
```diff
@@ -33,7 +33,7 @@ def serialization_scope_name=(scope_name)
       @adapter ||= ActiveModelSerializers::Adapter.create(serializer_instance, adapter_opts)
     end
-    alias adapter_instance adapter
+    alias_method :adapter_instance, :adapter
```
:exclamation: revert, revert!

bf4 commented on lib/active_model_serializers/adapter/base.rb:
```diff
@@ -1,6 +1,9 @@
 module ActiveModelSerializers
   module Adapter
     class Base
+      include ActiveModelSerializers::KeyTransform
```
why a module?

NullVoxPopuli replied:
what are you thinking for those set of methods?

bf4 replied:
`include` can now be removed as we're calling the function on KeyTransform directly

bf4 replied:
and maybe alias it `KeyTransform = ActiveModelSerializers::KeyTransform`

bf4 commented on lib/active_model_serializers/key_transform.rb:
```diff
@@ -0,0 +1,43 @@
+module ActiveModelSerializers
+  module KeyTransform
```
perhaps `module_function` here to allow public calling on the module but private methods on inclusion. I like me some utility function modules

NullVoxPopuli replied:
It used to be in this file. not sure what happened to it.

bf4 commented:
@remear changeset should include `config.jsonapi_resource_type = :plural` right?

bf4 commented on lib/active_model_serializers/key_transform.rb:
```diff
@@ -0,0 +1,43 @@
+    # @see: active_support/inflector#camelize
+    def camel(hash)
+      hash.deep_transform_keys! { |key| key.to_s.camelize.to_sym }
+    end
```
I think here you might mean to use a link like `# @see {https://github.com/rails/rails/blob/v5.0.0.beta3/activesupport/lib/active_support/inflector/methods.rb#L66-L76 ActiveSupport::Inflector.camelize}`

NullVoxPopuli replied:
I didn't know you could use that syntax for an actual link -- legit

remear commented:
I'm not sure how `config.jsonapi_resource_type = :plural` relates to these changes. Could you explain further?

bf4 commented on lib/active_model_serializers/key_transform.rb:
```diff
@@ -0,0 +1,43 @@
+
+    def self.method_missing(m, *args, &block)
+      warn "\"#{m}\" is not a defined key transform. Defaulting to :unaltered."
+      send(:unaltered, *args, &block)
```
this is super dangerous to eat a method missing entirely.  I think the proper behavior here is to raise.  I mean, if I say `KeyTransform.micky_mouse` it should fail, right?

NullVoxPopuli replied:
Do you mean you don't want to silently fail with a warning and default behavior? and would rather have it barf?

remear replied:
It just seemed nicer to just give back what was supplied and warn if they called an undefined method.

bf4 replied:
@remear @NullVoxPopuli So, before I asked:

> If I say `KeyTransform.micky_mouse` it should fail, right?

Here's a better example, I think:

`KeyTransform.caml`.  Person mistyped `camel` and now instead of a NoMethodError they're getting back `:unaltered`.  That's surprising.

I feel pretty strongly this should fail fast.  The warning message is good, but should be a failure message.  

Also, since you're defining the method missing on the module, are you also eating up all method missing calls on the `Adapter::Base`.  Just seems like asking for trouble with only benefit is a silent failure with a warning somewhere in the logs.

NullVoxPopuli replied:
oooo, that's a good point. that is a bad idea then.

bf4 commented on lib/active_model_serializers/serialization_context.rb:
```diff
@@ -4,13 +4,14 @@ class << self
     end
 
-    attr_reader :request_url, :query_parameters
+    attr_reader :request_url, :query_parameters, :key_case
```
why `key_case` and not `key_transform`? or `key_format`?

remear replied:
It should be key_transform now that we're using this KeyTransform interface.

bf4 commented:
@remear `config.jsonapi_resource_type = :plural` isn't that key casing concern?

remear commented:
I think it can be considered a key transformation concern, not a key casing one. The same goes with the change for removing `?` from keys. I would like to see things of this nature make their way into key transformation over time. This PR has been focused specifically on key casing but proposes the beginnings of the interface capable of handling all the key transformation. Do we want this PR scope to increase to accommodate the larger picture, or do we prefer to iterate over a few more PRs?

NullVoxPopuli commented:
My vote is for iterating over a few PRs.

NullVoxPopuli commented:
@remear can you rebase?

NullVoxPopuli commented on docs/general/key_transform.md:
```diff
@@ -0,0 +1,34 @@
+
+##### Adapter default
+
+Each adapter has a default key transform configured:
```
nice, I was going to ask about that

NullVoxPopuli commented:
LGTM :+1:

bf4 commented on docs/general/configuration_options.md:
```diff
@@ -2,26 +2,84 @@
+- `:camel` - ExampleKey
+- `:camel_lower` - exampleKey
+- `:dashed` - example-key
+- `:unaltered` - the original, unaltered key
```
missing default. perhaps intentional, but the others have it so, should mention something

remear replied:
It was intentional because it's mentioned elsewhere, but a mention would be good.

bf4 replied:
ok, I see that the missing default option is `nil: use the adapter default`

bf4 commented on docs/general/configuration_options.md:
```diff
@@ -2,26 +2,84 @@
+
+##### jsonapi_include_toplevel_object
+
+Include a [top level JSON API member](http://jsonapi.org/format/#document-jsonapi-object)
```
as long as we're here `s/JSON API/jsonapi`

remear replied:
Are these docs intended to refer to JSON API the spec? The adapter? Something else? Isn't the spec officially named `JSON API`?

bf4 replied:
Yeah, but the member name is `jsonapi`

bf4 commented on docs/general/key_transform.md:
```diff
@@ -0,0 +1,34 @@
+Each adapter has a default key transform configured:
+
+- `Json` - `:unaltered`
+- `JsonApi` - `:dashed`
```
`Attributes`? 

`Null`? lulz

bf4 replied:
I like calling dashed `kebab-case` like ember does... maybe I should make a PR to the ActiveSupport::Inflector to support that...

remear replied:
As far as I can tell, those are marked as deprecated, thus, on their way out.

bf4 replied:
good call.

bf4 commented on lib/active_model/serializer/configuration.rb:
```diff
@@ -26,6 +26,7 @@ def config.array_serializer
         # Make JSON API top-level jsonapi member opt-in
         # ref: http://jsonapi.org/format/#document-top-level
         config.jsonapi_include_toplevel_object = false
+        config.key_transform = nil
```
should be `:unaltered` no?

bf4 replied:
Then there's no need for `default_key_transform`

remear replied:
Seems like that would preclude the ability to have a per-adapter default.

bf4 replied:
oh n/m, that let's you configure per adapter. but this raise a point..

if this value is set to anything truthy, it will affect all adapters.  Is that the intention?

remear replied:
For now. Did you have something else in mind?

bf4 replied:
Just we should document that. Could be surprising. Or make it a required method in every adapter by just not defining it in the Base

bf4 commented on lib/active_model_serializers/adapter/json_api.rb:
```diff
@@ -37,15 +37,20 @@ def initialize(serializer, options = {})
+                   else
+                     failure_document
+                   end
+        transform_key_casing!(document, options[:serialization_context])
```
seeing as we're inheriting from a common base, I'm thinking we should just pass up the options?

remear replied:
Maybe. Right now it only needs the serialization context.

bf4 commented on lib/active_model_serializers/key_transform.rb:
```diff
@@ -0,0 +1,38 @@
+      hash
+    end
+
+    module_function :camel, :camel_lower, :dashed, :unaltered
```
Might as well just put the `module_function` as the top scope of the Module. There doesn't seem to be any reason to exclude methods

bf4 commented:
@remear :100: we need to add a controller test.  We've gotten in trouble before when not writing those :)

bf4 commented on lib/active_model_serializers/key_transform.rb:
```diff
@@ -0,0 +1,38 @@
+    #    "some_key" => "SomeKey",
+    # @see {https://github.com/rails/rails/blob/master/activesupport/lib/active_support/inflector/methods.rb#L66-L76 ActiveSupport::Inflector.camelize}
+    def camel(hash)
+      hash.deep_transform_keys! { |key| key.to_s.camelize.to_sym }
```
at top of file we should add `require 'active_support/core_ext/hash/keys'` to ensure the extension is loaded.

bf4 commented on docs/general/configuration_options.md:
```diff
@@ -2,26 +2,84 @@
+
+##### key_transform
+
+The [key transform](key_transform.md) to use.
```
`config.key_transform` is a global override of the adapter default.  Adapters still prefer the render option `:key_transform` over this setting

groyoh commented on docs/general/configuration_options.md:
```diff
@@ -2,26 +2,93 @@
+
+##### jsonapi_resource_type
+
+Set the `type` attributes of resources to singular or plural.
```
> Set the `type` attributes

Using `attributes` here seems inappropriate. What about something like this:
```md
Sets whether the [type](http://jsonapi.org/format/#document-resource-identifier-objects) of the JSON API resource should be `singularized` or `pluralized` when it is not [explicitly specified by the serializer](https://github.com/rails-api/active_model_serializers/blob/master/docs/general/serializers.md#type):
```

remear replied:
It could use more clarification. I didn't go out of my way to refactor the content here.

This is already in the JSON API section so the copy should read within that context.

```
Sets whether the [type](http://jsonapi.org/format/#document-resource-identifier-objects) of the resource should be `singularized` or `pluralized` when it is not [explicitly specified by the serializer](https://github.com/rails-api/active_model_serializers/blob/master/docs/general/serializers.md#type)
```

groyoh replied:
Sounds good.

rthbound commented:
@bf4, @remear 
It looks like there's some problems with this... The changes in [this branch](https://github.com/rails-api/active_model_serializers/compare/master...rthbound:bug-in-1574) should highlight what I'm talking about.

Basically, we're transforming keys only... `bad_comments` where it would appear in keys is properly transformed, however `bad_comments` in values (e.g. `type: "bad_comments"`) is not transformed.

This also suggests the action controller tests written for this were incomplete.

remear commented:
It doesn't cover everything yet. For example, [this](https://github.com/rails-api/active_model_serializers/blob/master/test/adapter/json_api/key_case_test.rb#L408), ideally, would have been transformed from `published_at` to `published-at`. Transforming all values is too heavy-handed, so more finesse is needed for these cases. We'll get it fixed up soon.

dgdosen commented:
Still - this was greatly needed.  Thanks!
